pkg/virtctl/scp/scp.go

```diff
@@ -0,0 +1,181 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package scp
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/povsister/scp"
+	"github.com/spf13/cobra"
+
+	"k8s.io/client-go/tools/clientcmd"
+
+	"kubevirt.io/kubevirt/pkg/virtctl/ssh"
+	"kubevirt.io/kubevirt/pkg/virtctl/templates"
+)
+
+func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+
+	c := &SCP{
+		clientConfig: clientConfig,
+		options:      ssh.DefaultSSHOptions(),
+		recursive:    false,
+		preserve:     false,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "scp (VM|VMI)",
+		Short:   "SCP files from/to a virtual machine instance.",
+		Example: usage(),
+		Args:    templates.ExactArgs("scp", 2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return c.Run(cmd, args)
+		},
+	}
+	cmd.Flags().BoolVarP(&c.recursive, "recursive", "r", c.recursive, "Recursively copy entire directories")
+	cmd.Flags().BoolVar(&c.preserve, "preserve", c.preserve, "Preserves modification times, access times, and modes from the original file.")
+	ssh.AddCommandlineArgs(cmd.Flags(), &c.options)
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+type SCP struct {
+	clientConfig clientcmd.ClientConfig
+	options      ssh.SSHOptions
+	recursive    bool
+	preserve     bool
+}
+
+func (o *SCP) Run(cmd *cobra.Command, args []string) error {
+	local, remote, toRemote, err := PrepareCommand(cmd, o.clientConfig, &o.options, args)
+	if err != nil {
+		return err
+	}
+
+	sshClient := ssh.NativeSSHConnection{
+		ClientConfig: o.clientConfig,
+		Options:      o.options,
+	}
+	client, err := sshClient.PrepareSSHClient(remote.Kind, remote.Namespace, remote.Name)
+	if err != nil {
+		return err
+	}
+	scpClient, err := scp.NewClientFromExistingSSH(client, &scp.ClientOption{})
+	if err != nil {
+		return err
+	}
+	if toRemote {
+		if o.recursive {
+			stat, err := os.Stat(local.Path)
+			if err != nil {
+				return fmt.Errorf("failed reading path %q: %v", local.Path, err)
+			}
+			if !stat.IsDir() {
+				return fmt.Errorf("local path %q is not a directory", local.Path)
+			}
+			err = scpClient.CopyDirToRemote(local.Path, remote.Path, &scp.DirTransferOption{PreserveProp: o.preserve})
```

Comment:
Regular `scp` would be okay with copying a file here, but I don't know how the scp lib handles that.

Reply:
It will fial with "error traversing dir".

Reply:
Improved.